File: 51-N-Queens.py
class Solution(object):
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        board = ['.'*n for _ in range(n)]
        rows = [False]*n
        cols = [False]*n
        diag1 = [False]*(2*n-1)
        diag2 = [False]*(2*n-1)
        res = []
        
        def isOccupied(r,c):
            if rows[r] or cols[c] or diag1[r+c] or diag2[r-c+n-1]:
                return True
            return False
        
        def updateBoard(r,c, put = True):
            if put:
                board[r] = board[r][:c] + 'Q' + board[r][c+1:]
            else:
                board[r] = board[r][:c] + '.' + board[r][c+1:]
            rows[r] = put
            cols[c] = put
            diag1[r+c] = put
            diag2[r-c+n-1] = put
            
        
        def dfs(r):
            if r == n:
                res.append(board[:])
                return
            
            for c in range(n):
                if isOccupied(r,c):
                    continue
                updateBoard(r,c)
                dfs(r+1)
                updateBoard(r,c, False)
                
        dfs(0)
        return res

File: JavaScript/51-N-Queens.js
/**
 * @param {number} n
 * @return {string[][]}
 */
var solveNQueens = function(n) {
    let row = new Array(n).fill(false), col = new Array(n).fill(false);
    let diag1 = new Array(2*n-1).fill(false), diag2 = new Array(2*n-1).fill(false);
    let res = [];
    let tmp = new Array(n).fill("");
    
    let dfs = (r) =>{
        if (r === n){
            res.push([...tmp]);
            return;
        }
        for (let c=0; c<n; c++){
            if (!row[r] && !col[c] && !diag1[r+c] && !diag2[r-c+n-1]){
                row[r] = true;
                col[c] = true;
                diag1[r+c] = true;
                diag2[r-c+n-1] = true;
                tmp[r] = ".".repeat(c) + "Q" + ".".repeat(n-1-c);
                dfs(r+1);
                tmp[r] = "";
                row[r] = false;
                col[c] = false;
                diag1[r+c] = false;
                diag2[r-c+n-1] = false;
            }
        }   
    }
    dfs(0);
    return res;
};